apiserver/facades/client/modelupgrader/upgrader.go

```diff
@@ -122,6 +122,41 @@ func (m *ModelUpgraderAPI) AbortModelUpgrade(arg params.ModelParam) error {
 	return st.AbortCurrentUpgrade()
 }
 
+// RequiredArchitectures returns all the architectures used by this model.
+func (m *ModelUpgraderAPI) RequiredArchitectures(arg params.ModelParam) (params.StringsResult, error) {
+	modelTag, err := names.ParseModelTag(arg.ModelTag)
+	if err != nil {
+		return params.StringsResult{}, errors.Trace(err)
+	}
+	st, err := m.statePool.Get(modelTag.Id())
+	if err != nil {
+		return params.StringsResult{}, errors.Trace(err)
+	}
+	defer st.Release()
+
+	machines, err := st.AllMachines()
+	if err != nil {
+		return params.StringsResult{}, errors.Trace(err)
+	}
+
+	archs := make(map[string]bool)
+	for _, m := range machines {
+		tools, err := m.AgentTools()
+		if errors.Is(err, errors.NotFound) {
+			continue
+		} else if err != nil {
+			return params.StringsResult{}, errors.Trace(err)
+		}
+		archs[tools.Version.Arch] = true
+	}
```

Comment:
Out of interest, why use this instead of instance data? Is this because we're not guaranteed to have instance data until it is reported back?

Reply:
I'm not sure, I just pulled it from the first place I could find. Essentially I need to know the architecture of the machines in the controller model, if you have a better idea, I would like to know.